### ISSUE TITLE
[HTML Report] Smaller hitbox area is covered by bigger hitbox area

### ISSUE THREAD

**Issue by Ridderby** [archimatetool/archi#520]
### Version of Archi, Operating System

Version 4.5.1, Linux 64 bit on Artix Linux. 
### Expected Behaviour

I have a "table of content"-view in the model where I have view links (drav a view element from the tree into the view) to help users navigate. 
In Archi I can double click on one and the view will open. 

In HTML reports generated by erlier versions one can single click on the element av nagivate to the view. Not sure which version though but working report was create 20190321 and I am fairly up to date. 

In a HTML report from current version nothing happens when I click on the element but if I select elements on the bottom part of the page the elements are listed with working links. 

### Actual Behaviour

I expect the view-links to work as before; click on them to navigate to the page. 

### Steps to Reproduce the Behaviour
1. Create page with view-links
2. Generate HTML report
3. Open HTML report
4. Click on the view-link

**Comment by Phillipus** [comment]
I tested this and it's working for me. Is there some other factor involved? Can you try on a different setup? Is it to do with screen resolution. Need more info to track this one down.

**Comment by Ridderby** [comment]
Hi, 

I found some more. 

I tend to frame my view-links using a Comment-box with descriptive tex or a header. The comment box frames the view-links a grouping. naturally I give the comment-box the lowest Z-order or it will hide everything else.  Now it seems like it is this box that gets the click in the report instead of the view-link. 

If I make view-links without a comment box around it works as usual. 

So to reproduce:
* Create view
* Create a comment box with a heading text in it (this is what I have)
* Drag a view from the model tree into the comment box
* Drag a second view from the tree into the view outside the comment box
* Generate HTML report
* Click on the view-link in the comment-box and it fails. 
* Click on the view-link outside the comment-box and it works.

My workaround was to make a small comment box with descriptive text above the row of view-links and it seems to work.

**Comment by Phillipus** [comment]
Yes, I see that. Probably because the hit box of the note overlaps the view link and is written to the HTML file after the view link. Will investigate at some point.

**Comment by jbsarrodie** [comment]
I faced this, and still face it in fact but I simply forgotten...

I discovered this almost 2 years ago without finding the reason. Time to really investigate.

One workaround is to create the view-link outside the comment-box, and then resize it so that it appears to be on top but is not really nested.

**Comment by Ridderby** [comment]
I took a look in a HTML report created in March this year, what ever relevant version I used by the time, and it seemed to worked fine. A report from the same model failed today (4.5.1). 
I never experienced this before but then again, I usually navigate the model in Archi. 

The frame is nice, the heading is important so I will be able to cope.

**Comment by Phillipus** [comment]
It's because selection of other graphical objects (Notes, Stickies, Canvas objects, Groups) was implemented in Archi 4.4. Before that they didn't have hitboxes.

**Comment by Ridderby** [comment]
Cool!

**Comment by Phillipus** [comment]
The Note is not a container so hitbox calculation is different.

You could use a Group figure (the grey ones), set the Border to "Rectangle" and change the colour so it looks like a Note.

**Comment by Phillipus** [comment]
@jbsarrodie I think it might be something to do with the order of processing children in `frame.stg`.

I can manually edit the generated html file and swap the order of `<area>` tags and it's OK.

**Comment by Phillipus** [comment]
It's the same problem for any smaller object that is in front of a non-container larger object.

**Comment by Phillipus** [comment]
In `frame.stg` a diagram's children are iterated through using the diagram model or object's getChildren() method. The order of children is the z-order.

For example - a smaller object is placed on top of a larger object in a diagram. The z-order of the diagram's children is then:

larger object
smaller object

Then when the hit areas are added in `frame.stg` (`expandAreasForChildren`) the area tags are created in that order like this:

```
<area shape="rect" coords="10,10,250,120" href="..." target="element" >  <-- larger object
<area shape="rect" coords="60,40,180,90" href="..."  target="element">    <-- smaller object
```

According to the HTML 4.01 spec:

"If two or more defined regions overlap, the region-defining element that appears earliest in the document takes precedence."

**Comment by Phillipus** [comment]
Don't know if it's possible to add a routine in the template to reverse the z-order of children?

**Comment by jbsarrodie** [comment]
> Don't know if it's possible to add a routine in the template to reverse the z-order of children?

Can't test for the moment, but you could try `<reverse(attr)>` (see https://github.com/antlr/stringtemplate4/blob/master/doc/cheatsheet.md#functions)

so on line 111, `^children:{c |` would become `^reverse(children):{c |`

**Comment by Phillipus** [comment]
@jbsarrodie Hot damn! That works!

Fix included in branch "html-report-area-fix".

We should test that this doesn't impact the existing behaviour.

**Comment by Phillipus** [comment]
I've tested the fix in various scenarios and it seems to be good. I'll merge to master and we can, hopefully, consider this one fixed.

**Comment by Phillipus** [comment]
BTW, setting the Opacity to 0 on a View Link means you can put it inside a note or another element for different visual effects

![View lInk](https://user-images.githubusercontent.com/600504/64100601-d0cc0a00-cd63-11e9-8f21-812caf3d8bf5.png)

**Comment by Phillipus** [comment]
Fixed in 4.6
